File: go/tags.go
package gencmu

import (
	"sort"
	"strings"
)

// A tag set (engine §1): tags in code point order, each strong or weak,
// interned so that an item can key on its id.
type tagset struct {
	id     int32
	names  []string
	strong []bool
	key    string
}

func (t *tagset) has(name string) (strong, ok bool) {
	i := sort.SearchStrings(t.names, name)
	if i < len(t.names) && t.names[i] == name {
		return t.strong[i], true
	}
	return false, false
}

func (t *tagset) toMap() map[string]bool {
	m := make(map[string]bool, len(t.names))
	for i, n := range t.names {
		m[n] = t.strong[i]
	}
	return m
}

type interner struct {
	byKey map[string]*tagset
	all   []*tagset
}

func newInterner() *interner {
	return &interner{byKey: map[string]*tagset{}}
}

func (in *interner) make(names []string, strong []bool) *tagset {
	var b strings.Builder
	for i, n := range names {
		b.WriteString(n)
		if strong[i] {
			b.WriteByte(1)
		} else {
			b.WriteByte(2)
		}
	}
	key := b.String()
	if t, ok := in.byKey[key]; ok {
		return t
	}
	t := &tagset{id: int32(len(in.all)), names: names, strong: strong, key: key}
	in.byKey[key] = t
	in.all = append(in.all, t)
	return t
}

func (in *interner) fromMap(m map[string]bool) *tagset {
	names := make([]string, 0, len(m))
	for n := range m {
		names = append(names, n)
	}
	sort.Strings(names)
	strong := make([]bool, len(names))
	for i, n := range names {
		strong[i] = m[n]
	}
	return in.make(names, strong)
}

func (in *interner) empty() *tagset { return in.make(nil, nil) }

func (in *interner) single(name string, strong bool) *tagset {
	return in.make([]string{name}, []bool{strong})
}

// union holds every tag of either, strong if strong in either.
func (in *interner) union(a, b *tagset) *tagset {
	if len(b.names) == 0 {
		return a
	}
	if len(a.names) == 0 {
		return b
	}
	var names []string
	var strong []bool
	i, j := 0, 0
	for i < len(a.names) || j < len(b.names) {
		switch {
		case j == len(b.names) || (i < len(a.names) && a.names[i] < b.names[j]):
			names, strong = append(names, a.names[i]), append(strong, a.strong[i])
			i++
		case i == len(a.names) || b.names[j] < a.names[i]:
			names, strong = append(names, b.names[j]), append(strong, b.strong[j])
			j++
		default:
			names, strong = append(names, a.names[i]), append(strong, a.strong[i] || b.strong[j])
			i++
			j++
		}
	}
	return in.make(names, strong)
}

// intersection holds the tags of a that are in b, with a's strength.
func (in *interner) intersection(a, b *tagset) *tagset {
	var names []string
	var strong []bool
	for i, n := range a.names {
		if _, ok := b.has(n); ok {
			names, strong = append(names, n), append(strong, a.strong[i])
		}
	}
	return in.make(names, strong)
}

// sameNames compares two sets by their tags alone, ignoring strength.
func sameNames(a, b *tagset) bool {
	if len(a.names) != len(b.names) {
		return false
	}
	for i := range a.names {
		if a.names[i] != b.names[i] {
			return false
		}
	}
	return true
}

// phonemeTag is the phoneme a tag /p/ names, a pause / / being a space: a
// phoneme tag is exactly three code points, the first and last / (§5).
func phonemeTag(tag string) (string, bool) {
	r := []rune(tag)
	if len(r) == 3 && r[0] == '/' && r[2] == '/' {
		return string(r[1]), true
	}
	return "", false
}

File: lib/rust/src/tags.rs
//! Tags and tag sets (engine §1): names interned to numbers, and sets of
//! (tag, strength) interned to numbers, so that items can compare them
//! cheaply.

use std::collections::BTreeMap;

use crate::fxhash::FxMap;

pub(crate) type TagId = u32;
pub(crate) type SetId = u32;

/// A tag set, sorted by tag id; `true` is strong.
pub(crate) type TagList = Vec<(TagId, bool)>;

#[derive(Debug, Default)]
pub(crate) struct Tags {
    names: Vec<String>,
    index: FxMap<String, TagId>,
    sets: Vec<TagList>,
    set_index: FxMap<TagList, SetId>,
}

impl Tags {
    pub(crate) fn new() -> Tags {
        let mut tags = Tags::default();
        tags.set(Vec::new());
        tags
    }

    pub(crate) fn tag(&mut self, name: &str) -> TagId {
        if let Some(&id) = self.index.get(name) {
            return id;
        }
        let id = self.names.len() as TagId;
        self.names.push(name.to_string());
        self.index.insert(name.to_string(), id);
        id
    }

    pub(crate) fn lookup(&self, name: &str) -> Option<TagId> {
        self.index.get(name).copied()
    }

    pub(crate) fn name(&self, id: TagId) -> &str {
        &self.names[id as usize]
    }

    /// Interns a set; the list must be sorted by tag id without repeats.
    pub(crate) fn set(&mut self, list: TagList) -> SetId {
        if let Some(&id) = self.set_index.get(&list) {
            return id;
        }
        let id = self.sets.len() as SetId;
        self.sets.push(list.clone());
        self.set_index.insert(list, id);
        id
    }

    pub(crate) fn list(&self, id: SetId) -> &TagList {
        &self.sets[id as usize]
    }

    pub(crate) fn contains(&self, set: SetId, tag: TagId) -> bool {
        self.sets[set as usize].binary_search_by_key(&tag, |&(id, _)| id).is_ok()
    }

    /// Builds a set from names and strengths; a tag given twice is strong
    /// if either is.
    pub(crate) fn set_of<'a>(&mut self, tags: impl IntoIterator<Item = (&'a str, bool)>) -> SetId {
        let mut list: TagList = tags.into_iter().map(|(name, strong)| (self.tag(name), strong)).collect();
        list.sort_unstable();
        list.dedup_by(|later, earlier| {
            if later.0 == earlier.0 {
                earlier.1 |= later.1;
                true
            } else {
                false
            }
        });
        self.set(list)
    }

    /// The set as a map from name to strength, in code point order.
    pub(crate) fn to_map(&self, set: SetId) -> BTreeMap<String, bool> {
        self.list(set).iter().map(|&(id, strong)| (self.name(id).to_string(), strong)).collect()
    }
}

/// The union of two sorted lists: every tag of either, strong if strong in
/// either.
pub(crate) fn union(left: &TagList, right: &TagList) -> TagList {
    let mut out = Vec::with_capacity(left.len() + right.len());
    let (mut i, mut j) = (0, 0);
    while i < left.len() && j < right.len() {
        match left[i].0.cmp(&right[j].0) {
            std::cmp::Ordering::Less => {
                out.push(left[i]);
                i += 1;
            }
            std::cmp::Ordering::Greater => {
                out.push(right[j]);
                j += 1;
            }
            std::cmp::Ordering::Equal => {
                out.push((left[i].0, left[i].1 || right[j].1));
                i += 1;
                j += 1;
            }
        }
    }
    out.extend_from_slice(&left[i..]);
    out.extend_from_slice(&right[j..]);
    out
}

/// The intersection: the tags of the first that are in the second, with
/// the first's strength.
pub(crate) fn intersection(left: &TagList, right: &TagList) -> TagList {
    left.iter().filter(|&&(id, _)| right.binary_search_by_key(&id, |&(other, _)| other).is_ok()).copied().collect()
}

/// Whether a tag is a phoneme tag `/p/`, exactly three code points, and if
/// so its phoneme `p`, or a space for the pause, `/./` (engine §5).
pub(crate) fn phoneme_of(tag: &str) -> Option<&str> {
    if tag == "/./" {
        Some(" ")
    } else if tag.chars().count() == 3 && tag.starts_with('/') && tag.ends_with('/') {
        Some(&tag[1..tag.len() - 1])
    } else {
        None
    }
}
